x-pack/plugins/monitoring/common/runtime_types.ts

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { fold } from 'fp-ts/lib/Either';
+import { pipe } from 'fp-ts/lib/pipeable';
+import { Context, Errors, IntersectionType, Type, UnionType, ValidationError } from 'io-ts';
+// eslint-disable-next-line @kbn/eslint/no-restricted-paths
+import type { RouteValidationFunction } from 'kibana/server';
+
+const getErrorPath = ([first, ...rest]: Context): string[] => {
+  if (typeof first === 'undefined') {
+    return [];
+  } else if (first.type instanceof IntersectionType) {
+    const [, ...next] = rest;
+    return getErrorPath(next);
+  } else if (first.type instanceof UnionType) {
+    const [, ...next] = rest;
+    return [first.key, ...getErrorPath(next)];
+  }
+
+  return [first.key, ...getErrorPath(rest)];
+};
+
+const getErrorType = ({ context }: ValidationError) =>
+  context[context.length - 1]?.type?.name ?? 'unknown';
+
+const formatError = (error: ValidationError) =>
+  error.message ??
+  `in ${getErrorPath(error.context).join('/')}: ${JSON.stringify(
```

Comment:
`JSON.stringify(value)` in a response got my XSS radar tingling here (stuff like https://pragmaticwebsecurity.com/articles/spasecurity/json-stringify-xss.html). I guess it's probably safe because it's inside of a non-parsable string. Got me wondering if we really want the whole error.value or if just some part of it is enough to help diagnose the problem.

Reply:
> Got me wondering if we really want the whole error.value or if just some part of it is enough to help diagnose the problem.

Honestly I think with io-ts and the `ValidationError` we get the whole `value` is quite valuable. It can be verbose, but usually that's helpful when you're dealing with io-ts problems.

> JSON.stringify(value) in a response got my XSS radar tingling here

Yeah, that's a good point. I *think* we're okay here, like you say it's inside of a non-parsable string. I don't feel confident saying there's 100% not an attack vector though. We've been using this in `infra` for a very long time though, so maybe we stick with it and double check with someone from (I guess) Kibana security?

Reply:
Do we understand the ways in which `error.value` is set? The only possible vector I see would be:

1. Some error produces a string with `<script>` tags which is somehow [a] included inside of `error.value`
2. We return that value in our API request [b]
3. We use the returned API result to generate some kind of error message in the UI, which includes and executes the vulnerable string [c]

So my questions are: 
[a] Can arbitrary errors create any `error.value` value that they want?
[b] Are these error strings returned in the API payload (vs just being logged somewhere for us)?
[c] Do we have access to this part of the returned API payload to use in the UI?

If even just [a] and [b] are "yes" I think we likely want to do some basic escaping of HTML chars in those values.

Reply:
I would tend to disagree here. The API should be allowed to return any kind of string in the values of the JSON object. It's agnostic to the fact that HTML even exists. If it was written to a log file or piped to some other process a different kind of escaping would be necessary, which we can't anticipate. Instead it's the responsibility of the renderer (e.g. our browser-side React code not to render markup that can cause injection). If we use `dangerouslySetInnerHTML` in our React code that's the place where alarm bells should ring.

Reply:
The `JSON.stringify` here is used to get a serialized representation of the value that failed the validation. It will be part of the value in `left()` of the `Either` or in the `message` of the `Error` if used in `decodeOrThrow()`. The function is pure and doesn't have a side-effect like writing to a page.

We could use a more basic conversion to string, but then we'd only see representations like `[Object object]` in the error message which would be a lot less useful and wouldn't be more or less safe as far as I can tell.